microsoft/iot-cardboard-js
# [Data History Explorer]: Error Handling

### Summary of changes 🔍 
_Note:_ This branch is off the [Telemetry branch](https://github.com/microsoft/iot-cardboard-js/pull/840), merge that one first for less file diffs.
- Created a standalone component **`DataHistoryErrorHandlingWrapper`** to handle ADX service specific errors and `TimeSeriesDatabaseConnectionFetchFailed` adapter error.
   - This component is consumed by `DataHistoryWidget` and `DataHistoryExplorer` now. 
- Added series level "noData" errors if there is no ADX data to display in the current view (either in Chart or Table)
   - For already numeric properties it might be due to the selected time range
   - For non numeric properties but which is set as "cast to numeric", the ADX still may return to null since with this toggle we add toDouble() in the ADX query for the Value column but which can still return null if the value is not castable to double (e.g. even though a boolean property is casted to numeric in setting, the toDouble('True') will return null since it is not castable to a number, but when a string property which stores a numeric value is casted to numeric toDouble('123') will return data like 123).
- Changed some translation for messages to better reflect the errors
- Here is the list of handled errors (I put the references into the DataHistoryErrorHandlingWrapper component):
   - **Cluster connection** (this happens when ADX request fails with status 0 which might be due to invalid cluster url, stopped cluster or network connectivity error) 
   - **Database connection** (this happens with specific ADX return error named '_BadRequest_EntityNotFound_' with status 400)
   - **Table or query error** (since table is part of the query payload in the request, and this happens with specific ADX return error named '_General_BadRequest_' with status 400)
   - Missing permission to query (this happens when ADX request fails with error code 403)
   - Additionally, I added to handle '_TimeSeriesDatabaseConnectionFetchFailed_' custom adapter error (this happens when `getTimeSeriesConnectionInformation` from AzureManagementAdapter fails to fetch the connection information of the selected ADT instance which might be due to control plane access or if there is no connection setup for the instance)

**To-do:** I am going to do content pass with @azimamansuri and @rinisbet 
![image](https://user-images.githubusercontent.com/45217314/215174511-eb29b285-24bd-4ab7-938b-1731385af180.png)
![image](https://user-images.githubusercontent.com/45217314/215174621-cc55cf17-fef8-4c07-8d6b-c3c1fa162eb4.png)
![image](https://user-images.githubusercontent.com/45217314/215174732-30844ce5-5d0b-447c-8d33-ce3ed8d24ea5.png)
![image](https://user-images.githubusercontent.com/45217314/215174777-fda42ab5-47c1-41ce-8a72-7fb32198fc21.png)
![image](https://user-images.githubusercontent.com/45217314/215174821-fc085a38-f287-4b92-b812-d7a0df78157f.png)


### Testing 🧪
- Check `DataHistoryErrorHandlingWrapper` individual stories, or
- Check Components/DataHistoryExplorer/Internal/TimeSeriesViewer/Errors stories or
- Components/DataHistoryExplorer/Modal/TimeSeriesConnectionInformationFetchFailed or
- Check the whole integration on Components/DataHistoryExplorer/Modal/ADT with actual data since MockAdapter always return a mock data currently.
- You can test some of the errors by:
   - Set a no data history enabled adt instance url in your secrets file
   - Stop your cluster that your adt instance connected
      -  It is hard to test database and table errors since you cannot change the name of those or stop the service but you can try to delete a data connection which I didn't try yet.
   - Try to select series which you think don't have data in selected time frame or data type would not return to numeric even if it is set as casted (see my comment above about those). 
   - Try to remove your permission to query from the cluster which is again hard to test because we all have Contributor and Owner assignments due to security group access. You can try to login the stories with a guest user account though. In that case you would probably see the _TimeSeriesDatabaseConnectionFetchFailed_ error though.

### Checklist ✔️
